pkg/ccl/backupccl/restore_planning.go

```diff
@@ -158,6 +159,237 @@ func synthesizePGTempSchema(
 	return synthesizedSchemaID, err
 }
 
+func bumpDescIDSeqKey(
+	ctx context.Context,
+	p sql.PlanHookState,
+	databasesByID map[descpb.ID]*dbdesc.Mutable,
+	schemasByID map[descpb.ID]*schemadesc.Mutable,
+	tablesByID map[descpb.ID]*tabledesc.Mutable,
+	typesByID map[descpb.ID]*typedesc.Mutable,
+) error {
+	// Include the database descriptors when calculating the max ID.
+	var maxDescIDInBackup int64
+	for _, table := range tablesByID {
+		if int64(table.ID) > maxDescIDInBackup {
+			maxDescIDInBackup = int64(table.ID)
+		}
+	}
+
+	for _, database := range databasesByID {
+		if int64(database.ID) > maxDescIDInBackup {
+			maxDescIDInBackup = int64(database.ID)
+		}
+	}
+
+	// Include the type descriptors when calculating the max ID.
+	for _, typ := range typesByID {
+		if int64(typ.ID) > maxDescIDInBackup {
+			maxDescIDInBackup = int64(typ.ID)
+		}
+	}
+
+	// Include the schema descriptors when calculating the max ID.
+	for _, sc := range schemasByID {
+		if int64(sc.ID) > maxDescIDInBackup {
+			maxDescIDInBackup = int64(sc.ID)
+		}
+	}
+
+	return p.ExecCfg().DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		v, err := txn.Get(ctx, p.ExecCfg().Codec.DescIDSequenceKey())
+		if err != nil {
+			return err
+		}
+		newValue := maxDescIDInBackup + 1
+		if newValue <= v.ValueInt() {
+			// This case may happen when restoring backups from older versions.
+			newValue = v.ValueInt() + 1
+		}
+		b := txn.NewBatch()
+		// N.B. This key is usually mutated using the Inc command. That
+		// command warns that if the key was every Put directly, Inc will
+		// return an error. This is only to ensure that the type of the key
+		// doesn't change. Here we just need to be very careful that we only
+		// write int64 values.
+		// The generator's value should be set to the value of the next ID
+		// to generate.
+		b.Put(p.ExecCfg().Codec.DescIDSequenceKey(), newValue)
+		return txn.Run(ctx, b)
+	})
+}
+
+func findConflictingSystemTables(
+	ctx context.Context,
+	p sql.PlanHookState,
+	databasesByID map[descpb.ID]*dbdesc.Mutable,
+	schemasByID map[descpb.ID]*schemadesc.Mutable,
+	tablesByID map[descpb.ID]*tabledesc.Mutable,
+	typesByID map[descpb.ID]*typedesc.Mutable,
+) (map[descpb.ID]catalog.TableDescriptor, error) {
+	systemTablesWithConflictingIDs := make(map[descpb.ID]catalog.TableDescriptor)
+	err := sql.DescsTxn(ctx, p.ExecCfg(), func(ctx context.Context, txn *kv.Txn, col *descs.Collection) error {
+		systemTableDescs, err := col.GetAllTableDescriptorsInDatabase(ctx, txn, systemschema.SystemDB.GetID())
+		if err != nil {
+			return err
+		}
+		for _, sysTableDesc := range systemTableDescs {
+			id := sysTableDesc.GetID()
+			if desc, ok := databasesByID[id]; ok {
+				if desc.GetParentID() != systemschema.SystemDB.GetID() {
```

Comment:
This should be GetID() instead of GetParentID() I think.

Reply:
Nice catch, fixed.